Add GraphQL mutation to merge identities

mafesan opened the pull request:
This PR adds the functionality to merge two unique identities, including their identities, enrollments and profiles. 

This functionality works the same than in the original version of SortingHat **with one exception**: in the original version this method did nothing when `fromUuid` and `toUuid` were equal. Now, when this happens it returns an error.

sduenas commented on tests/test_schema.py:
```diff
@@ -1798,3 +1803,196 @@ def test_non_existing_enrollments(self):
+        self.assertEqual(uuid, 'caa5ebfe833371e23f0a3566f2b7ef4a984c4fed')
+
+        # Check database objects
+
```
Remove this line

mafesan replied:
Done

sduenas commented on tests/test_schema.py:
```diff
@@ -1798,3 +1803,196 @@ def test_non_existing_enrollments(self):
+        uidentity_db = UniqueIdentity.objects.get(uuid='caa5ebfe833371e23f0a3566f2b7ef4a984c4fed')
+
+        self.assertIsInstance(uidentity_db, UniqueIdentity)
+
```
Remove this empty line.

mafesan replied:
Done

mafesan commented on sortinghat/core/api.py:
```diff
@@ -503,3 +504,107 @@ def withdraw(uuid, organization, from_date=None, to_date=None):
+    """
+    Merge one unique identity into another.
+
+    Use this function to join 'from_uuid' unique identity into
```
Done, I had copied part of the text to follow the format and I wasn't being consistent formatting these references.

mafesan commented on sortinghat/core/api.py:
```diff
@@ -503,3 +504,107 @@ def withdraw(uuid, organization, from_date=None, to_date=None):
+
+    uidentity = to_uid
+    for identity in identities:
+        uidentity = move_identity_db(identity, to_uid)
```
Yes. This was because first I didn't want to modify the value of `to_uid` variable, but after your comment I think it does not make sense to have `uidentity` as a variable. Now, only `to_uid` is used.

mafesan commented on sortinghat/core/api.py:
```diff
@@ -503,3 +504,107 @@ def withdraw(uuid, organization, from_date=None, to_date=None):
+
+
+@django.db.transaction.atomic
+def merge_identities(from_uuid, to_uuid):
```
I totally agree. I saw in the original version of SortingHat this procedure was included in one single method, but here it became too long. I have created to private methods inside the outer method as you suggested: `_merge_enrollments` and `_merge_profiles`.

mafesan commented on tests/test_api.py:
```diff
@@ -1482,3 +1485,303 @@ def test_non_existing_enrollment(self):
+
+        # Tests
+        self.assertIsInstance(uidentity, UniqueIdentity)
+
```
Done

mafesan commented on tests/test_api.py:
```diff
@@ -1482,3 +1485,303 @@ def test_non_existing_enrollment(self):
+        self.assertEqual(profile.gender_acc, 75)
+        self.assertEqual(profile.is_bot, True)
+        self.assertEqual(profile.country_id, 'US')
+
```
Done.

mafesan commented on tests/test_api.py:
```diff
@@ -1482,3 +1485,303 @@ def test_non_existing_enrollment(self):
+        self.assertEqual(rol2.end, datetime.datetime(2100, 1, 1, tzinfo=UTC))
+
+    def test_duplicate_enrollments(self):
+        """Check whether it merges two unique identities having duplicate enrollments"""
```
You are right. Now we have three different tests:
* `test_duplicate_enrollments`: It tests the case where the merged identities had exactly the same enrollment (Organization, start date and end date).
* `test_overlapping_enrollments`: It tests the case where the merged identities had different enrollments, with the same organization but with overlapping periods.
* `test_overlapping_enrollments_different_orgs`: It tests the case where the merged identities had different enrollments, including different organizations and overlapping periods.

mafesan commented on tests/test_api.py:
```diff
@@ -1482,3 +1485,303 @@ def test_non_existing_enrollment(self):
+        self.assertEqual(rol3.end, datetime.datetime(2100, 1, 1, tzinfo=UTC))
+
+    def test_overlapping_enrollments(self):
+        """Check whether it merges two unique identities, merging their enrollments with overlapping periods"""
```
I agree, please check my comment above about how I solved it: https://github.com/chaoss/grimoirelab-sortinghat/pull/212#discussion_r321815569.

mafesan commented on tests/test_api.py:
```diff
@@ -1482,3 +1485,303 @@ def test_non_existing_enrollment(self):
+        self.assertLessEqual(before_merge_dt, uid.last_modified)
+        self.assertGreaterEqual(after_merge_dt, uid.last_modified)
+
+        # Not merged identity were not modified
```
I don't fully understand this comment :sweat_smile:.
The dates of the merged identity are checked in the previous lines, do you miss any additional check? Or am I doing this check wrong?

mafesan replied:
This is clarified now. The unique identity which has been merged contains 4 identities. The dates from the 2 moved identities should change, whilst dates from the other 2 should remain the same. I'm working to fix this.

mafesan commented:
Thank you for the review, @sduenas. I have answered and addressed your comments, please check them again.

There is only one comment where I need some additional clarification, please check https://github.com/chaoss/grimoirelab-sortinghat/pull/212#discussion_r321241252. 

Thanks!
